Refactoring of work packages table

oliverguenther opened the pull request:
### ToDos

**Basic rendering**
- [x] Create basic class-based separation of table/row/cell renderers.
- [x] Delegation to `DisplayField` for rendering read-only state cells
- [x] Grouping
- [x] Removing templated rendering whenever possible
  - [x] display-attr directive
  - [x] id-field template
  - [x] `ui-state` directives in work package links.

**State handling**
- [x] Row selection state and manipulation
- [x] Edit form replacement as a state (aka `Rows currently being edited`).
- [x] Query is not always causing a correctly refreshed table (e.g, group by column after loaded).
- [x] Fix wiring of wp-list.controller and new table implementation.

**Other**
- [x] Context menu handling
  - [x] Fix keypress open / focussing through shift+alt+f10
- [x] Refactor other context menu to use global ngContextMenu service
- [x] Replace angular-busy with some other loading indicator outside angular
- [x] Fix wiring of PaginationService and query
- [x] Global Table sums (does not include row-group sums)
- [X] Adapt tests after simplification of display/edit field structure.
- [x] Get rid of cg-busy

**Bugs**

- [x] *When grouping by type*, and switching the type to another with a required CF, the column expansion of this CF redraws the table, which overrides the group since the type is switched.

**Out of scope**
- Re-add Inline create row

- Group header sums

- Edit fields compatibility

- [Sticky headers](https://github.com/opf/openproject/pull/5117#issuecomment-269121885)

- Visualization of hierarchies

## Previous discussion


This PR sketches my ideas of the work packages table refactoring with the following goals in mind:

**Preparation for features**

* Expanding parent / child relations ([#24455](https://community.openproject.com/projects/openproject/work_packages/24455))
* Expanding work package relations ([#24454](https://community.openproject.com/projects/openproject/work_packages/details/24454))
*  Drag & Drop of rows
  * reordering work packages arbitrarily
  * within attribute groups
* Potential additions to watch for
  * Resizable columns(?)
  * Tooltips in work package table

**Performance**
 
* Faster initial rendering of the table
* Faster refreshing of the table upon one-dimensional changes
  * Removal / Addition of columns
  * Removal / Addition of rows
  * Refreshing of a single row
* Angular-less implementation in the view-only mode of the table
* Global event handlers for actions on the table (Click to edit, Drag to Move, Right click for context menu)

**Accessibility**

* The hierarchy of the work package table should be accessible https://community.openproject.com/work_packages/23471

### Implementation approach

We already heavily use Rx.js for communicating changes to work packages and other resources. Thus, we already know *when* a work package is updated through its observable state.

#### Table state

Applying this pattern to the table results in a state for the current table, possibly divided into separate states for the current set of columns and visible work package rows.

#### Table renderers

I suggest to replace the templated ng-repeat structure of the table (mainly located within `wp-table.directive.html`) with a 3-step renderer for the table.

1st step: **Static initial page setup**

* Render the empty table skeleton and hook the renderers to it.
* Initialize an empty row object renderer
* Initialize an empty column renderer
* Fulfill the request to receive all work packages and associated schemas
* Receive query result of N rows, C columns
* Create column renderer of appropriate type
* Update page information from query result

2nd step: **Row rendering creation**

* Iterate through the received work packages |N|
  * Create and link work package rows from the loaded resources
  * Prepend group sum row when necessary
  * Prepend grouped row when necessary
  * Apply cell renderer to the work package row for the given type (|C| instances of DisplayField)

3rd step: **Additional information**

* Render sums schema (if necessary)
* Render inline-create row
* Process additional requests
  * Relations

#### Expansion of a row

The expansion of a row is handled by a separate renderer that receives as input a row from which elements are expanded as well as a facade that returns a set of rows to insert.

They alter the global representation of work package rows N.

#### Updating all rows (Refreshing the entire view)

Event: Current refreshing after work package change
Event: Page navigation / Property change
Event: Filter change	

Receive work packages result of N rows for the current query object.
Empty TBODY and re-execute steps 2 and 3.

Potential later optimization:

* Remove all Rows > |N|
* Empty existing TR and re-use.


#### Updating columns

Event: Column object C is updated in state.

All these operations iterate over the set of work package rows (N).
Initially, refresh the entire view.

Potential later optimization:
* Insert: Create a new column representation and splice the DOM to insert the new value
* Remove: Drop column header and cells
* Move: Swap DOM columns


#### Updating a single row

Event: Work Package state of <ID> is updated.

All operations of a work package function on a single work package row, identified by its unique identifier.

The row is always re-rendered fully unless it is dirty (being edited).

#### Event handling

I did not make explicit tests of the impact of current event listeners on the table performance. However, since we can delegate events, I suggest that we try to globalize all standard events and delegate the event to specific handlers that decide what result should come of it.

* Click on tbody row (selection)
* Click on field (Activate edit if possible)
* Click on details button
* (Future: Drag)

### Alternatives

None considered.

oliverguenther committed:
New table class with delegated builders and form handlers

lindenthal commented:
Hi @oliverguenther, 
I just had a look at the sticky header feature which allows for sticky headers using CSS only. This feature is available in FF and will be availabe (again) in Chrome 56. Does this help us to improve performance of the work packages table when compared with a JS solution?

https://developers.google.com/web/updates/2016/12/position-sticky

Best,
Niels

oliverguenther committed:
Working row selection

oliverguenther committed:
Details link click handling

oliverguenther committed:
Extract external template rendering process

oliverguenther committed:
Editing functionality WIP

oliverguenther committed:
Basic table edit field and error handling functionality

oliverguenther committed:
Introduce separate click/keypress handlers

oliverguenther committed:
Move to columns state

oliverguenther committed:
Allow listening to changes on multi state members

oliverguenther committed:
WIP column movements

oliverguenther committed:
Use wpColumnsService in column modal

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ae10f79cf13dac4e0f7bd181e5045c84779b4849) (deployed commit: 4b1179c7fb2b51d14282cb6add36adb4ab6c841c).

oliverguenther committed:
Refactor click and enter handlers into one

oliverguenther committed:
Avoid empty draw

oliverguenther committed:
Rename parent multi state

oliverguenther committed:
Use focusedWorkPackage for navigation

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/55cbeb752d93995eaedc63bd8018b0524e684e24) (deployed commit: 4b1179c7fb2b51d14282cb6add36adb4ab6c841c).

oliverguenther committed:
Fix single view wp

oliverguenther committed:
Add actual changed value to MultiState#changed

oliverguenther committed:
Remove workPackageId number|string dualism

States always expect strings, so we should treat the ids as such.

oliverguenther committed:
Avoid firing row update when state was cleared

oliverguenther committed:
Restore ng-required

oliverguenther committed:
Extract state mutation parts of table implementation as transformers

oliverguenther committed:
Group state handling

oliverguenther committed:
Avoid window resizing watcher

oliverguenther committed:
Fix package.json after dev rebase

oliverguenther committed:
Output correct link in ui-state-link-builder

oliverguenther committed:
Refactor context menus using global service

oliverguenther committed:
Position dropdown after rendering complete

oliverguenther committed:
Fix global mousetrap exposer

oliverguenther committed:
Add empty rows renderer

oliverguenther committed:
Fix wp selector in context menu spec

oliverguenther committed:
Avoid keeping a reference to an old row element

oliverguenther committed:
Let has-dropdown-menu directive reuse global contextmenu service

oliverguenther committed:
Remove $broadcasts for dropdowns/context menus

oliverguenther committed:
Add TableWorkPackageField due to incompatible edit field syntax

oliverguenther committed:
Implement new non-digest based loadingIndicator

oliverguenther committed:
Replace information in WorkPackageTableService with a metadata state

oliverguenther committed:
Remove unused WPTableService and WPTableHelper

oliverguenther committed:
Add WP Table metadata service

oliverguenther committed:
Subscribe to metadata for updating pagination

oliverguenther committed:
Update more intelligently when a row is being edited

Fixes the following bug:
When grouping by type, and switching the type to another with a required CF, the column expansion of this CF redraws the table, which overrides the group since the type is switched.

oliverguenther committed:
Work around schema loading in wp-list controller

oliverguenther committed:
Fix grouping by index when losing a group after refreshing

oliverguenther committed:
Update columns in query when changing state

oliverguenther committed:
Restore sums tfoot

oliverguenther committed:
Respect a field's hidden state

oliverguenther committed:
Fix specs

oliverguenther committed:
Fix keyboard context menu handler and focus resets

oliverguenther committed:
Fix specs

oliverguenther committed:
Fix focusing work package on details deep link

oliverguenther committed:
Fix group header spec

oliverguenther committed:
Fix row transformer listening on empty/null rows

oliverguenther committed:
Fix npm tests

oliverguenther committed:
Fix more specs

oliverguenther committed:
Compute unusedColumns correctly

oliverguenther committed:
Fix more tests

oliverguenther committed:
Add debug logging and fix codeclimate issues

oliverguenther committed:
Remove angular-busy

oliverguenther committed:
Fix read-only styles

oliverguenther committed:
Remove manualRenderer property

No longer required since fields can decide themselves how to render.

oliverguenther committed:
Fix rendering description html

oliverguenther committed:
observeUntil to scope rx state listeners

oliverguenther committed:
observeUntil to scope rx state listeners

oliverguenther committed:
Avoid observeUntil in wp-list ctrl

oliverguenther commented:
Costs PR: https://github.com/finnlabs/openproject-costs/pull/276

oliverguenther commented:
@romanroe Codeclimate reports an error despite having no issues, I've refreshed the branch over there, but so far no luck. Otherwise ready to review from my side.

romanroe commented on frontend/app/components/common/loading-indicator/loading-indicator.service.ts:
```diff
@@ -26,12 +26,62 @@
+  }
+
+  public stop() {
+    this.indicator.find(indicatorBackgroundSelector).remove();
```
Wouldn't `this.element.remove()` work?

oliverguenther replied:
It does, done.

romanroe commented on frontend/app/components/context-menus/column-context-menu/column-context-menu.controller.ts:
```diff
@@ -0,0 +1,147 @@
+
+  $scope.I18n = I18n;
+
+  $scope.$watch('column', function () {
```
Still use of `scope` despite the `States...` ?

oliverguenther replied:
That's due to the query service and table columns still watching on `scope.columns`. I agree that we should move that to the states, but that involves quite a lot more than the table and also reaches into frontend work on the query following Jens' API redesign.

-> Fully agreed, but let's address this in a separate PR.

romanroe commented on frontend/app/components/context-menus/column-context-menu/column-context-menu.controller.ts:
```diff
@@ -0,0 +1,147 @@
+      return ((currentState === undefined) ? state : currentState && !state);
+    };
+
+    switch (feature) {
```
Woops! No `break`s !!!

romanroe commented on frontend/app/components/context-menus/context-menu.service.ts:
```diff
@@ -0,0 +1,117 @@
+
+interface ContextMenu {
+  active();
+  close(disableFocus);
```
- return type
- Argument type declaration would be nice

romanroe replied:
- return type

romanroe commented on frontend/app/components/context-menus/context-menu.service.ts:
```diff
@@ -0,0 +1,117 @@
+// ++
+
+interface ContextMenu {
+  active();
```
- return type

romanroe commented on frontend/app/components/context-menus/context-menu.service.ts:
```diff
@@ -0,0 +1,117 @@
+interface ContextMenu {
+  active();
+  close(disableFocus);
+  open();
```
- return type

romanroe commented on frontend/app/components/context-menus/context-menu.service.ts:
```diff
@@ -0,0 +1,117 @@
+}
+
+export class ContextMenuService {
+  private _active_menu:ContextMenu|null;
```
Why leading `_` ? Not our coding convention, or is it?

romanroe commented on frontend/app/components/context-menus/context-menu.service.ts:
```diff
@@ -0,0 +1,117 @@
+export class ContextMenuService {
+  private _active_menu:ContextMenu|null;
+
+  constructor(public $window, public $injector, public $q, public $timeout, public $rootScope) {
```
missing parameter type annotations

romanroe commented on frontend/app/components/context-menus/context-menu.service.ts:
```diff
@@ -0,0 +1,117 @@
+    }
+  }
+
+  public activate(contextMenuName, event:JQueryEventObject, locals, positionArgs?:any) {
```
1. no annotation
2. good
3. `any` annotation

At least option 3., better always add a valid type

romanroe commented on frontend/app/components/context-menus/context-menu.service.ts:
```diff
@@ -0,0 +1,117 @@
+      contextMenu.menuElement = menuElement;
+      this._active_menu = contextMenu;
+      (menuElement as any).trap();
+      menuElement.on('click', function (e) {
```
replace `function` with arrow function. `function` is prone to future bugs, when some introduces a line with `this`

romanroe commented on frontend/app/components/context-menus/context-menu.service.ts:
```diff
@@ -0,0 +1,117 @@
+    return deferred.promise;
+  }
+
+  public reposition(event:JQueryEventObject, positionArgs?) {
```
missing parameter type

romanroe commented on frontend/app/components/context-menus/has-dropdown-menu/has-dropdown-menu-directive.ts:
```diff
@@ -0,0 +1,88 @@
+function hasDropdownMenu(contextMenu:ContextMenuService, FocusHelper) {
+  return {
+    restrict: 'A',
+    link: function (scope, element, attrs) {
```
missing parameter types

romanroe commented on frontend/app/components/context-menus/has-dropdown-menu/has-dropdown-menu-directive.ts:
```diff
@@ -0,0 +1,88 @@
+    restrict: 'A',
+    link: function (scope, element, attrs) {
+      var menuName = attrs['target'],
+        locals = {},
```
ugly (IMHO) variable declaration

romanroe commented on frontend/app/components/context-menus/has-dropdown-menu/has-dropdown-menu-directive.ts:
```diff
@@ -0,0 +1,88 @@
+        collisionContainer = attrs['collisionContainer'],
+        triggerOnEvent = (attrs['triggerOnEvent'] || 'click') + '.dropdown.openproject';
+
+      function open(event) {
```
missing parameter types

romanroe commented on frontend/app/components/context-menus/has-dropdown-menu/has-dropdown-menu-directive.ts:
```diff
@@ -0,0 +1,88 @@
+        });
+      }
+
+      function close(ignoreFocusOpener) {
```
missing parameter types

romanroe commented on frontend/app/components/context-menus/settings-menu/settings-menu.controller.ts:
```diff
@@ -27,6 +26,7 @@
 import {opWorkPackagesModule} from '../../../angular-modules';
+import {ContextMenuService} from '../context-menu.service';
 
 function SettingsDropdownMenuController($scope,
```
missing parameter types

romanroe commented on frontend/app/components/filters/query/query-service.service.ts:
```diff
@@ -188,8 +181,11 @@ function QueryService($rootScope,
     },
 
     selectUnusedColumns: function(columns) {
```
missing parameter types

romanroe commented on frontend/app/components/modals/grouping-modal/grouping-modal.service.html:
```diff
@@ -27,7 +27,7 @@
                 id="selected_columns_new"
                 class="form--select"
-                ng-options="column.name as column.title for column in vm.groupableColumns | orderBy: 'title'">
+                ng-options="column.name as column.title for column in vm.groupableColumns() | orderBy: 'title'">
```
Better name could be `getGroupableColumns`

romanroe commented on frontend/app/components/routing/wp-list/wp-list.controller.ts:
```diff
@@ -124,49 +132,42 @@ function WorkPackagesListController($scope,
 
-    // query data
-    QueryService.setTotalEntries(json.resource.total);
+    // Set metadata from results
```
`const` ? (lots of places)

romanroe commented on frontend/app/components/work-packages/work-package-comment/wp-comment-field.module.ts:
```diff
@@ -32,7 +32,7 @@ export class WorkPackageCommentField extends WikiTextareaEditField {
   public fieldVal = { raw: '' };
 
-  constructor(workPackage, protected I18n) {
+  constructor(workPackage, public I18n) {
```
missing parameter types

romanroe commented on frontend/app/components/work-packages/wp-display-attr/wp-display-attr.directive.ts:
```diff
@@ -89,25 +90,28 @@ export class WorkPackageDisplayAttributeController {
   protected updateAttribute(wp) {
     this.workPackage = wp;
-    this.field = <DisplayField>this.wpDisplayField.getField(this.workPackage, this.attribute, this.schema[this.attribute]);
+    this.field = <DisplayField> this.wpDisplayField.getField(this.workPackage, this.attribute, this.schema[this.attribute]);
```
Please use `... as DisplayField` to avoid syntax confusion with TSX files

romanroe commented on frontend/app/components/wp-create/wp-create.controller.ts:
```diff
@@ -92,6 +102,12 @@ export class WorkPackageCreateController {
       .catch(error => this.wpNotificationsService.handleErrorResponse(error));
   }
 
+  protected $inject(...args:string[]) {
```
maybe refactor this method to a "util" module? What about `injectorBridge`?

oliverguenther replied:
Will not work due to WPCopyController being a angular controller subclass, and injectorBridge would only inject the dependencies of the subclass due to the `Constructor.$inject = [..]` notation.

Let's think of some util module that works in all cases and replace all other methods with it in a following PR. Could you think about a method that supports all these cases?

romanroe commented on frontend/app/components/wp-display/field-types/wp-display-date-field.module.ts:
```diff
@@ -26,15 +26,24 @@
 export class DateDisplayField extends DisplayField {
 
-  isManualRenderer = true;
+  public WorkPackagesHelper;
```
missing parameter types

romanroe commented on frontend/app/components/wp-display/field-types/wp-display-id-field.module.ts:
```diff
@@ -53,7 +59,23 @@ export class IdDisplayField extends DisplayField {
     }
   }
 
+  public render(element: HTMLElement, displayText): void {
```
missing parameter types

romanroe commented on frontend/app/components/wp-display/field-types/wp-display-spent-time-field.module.ts:
```diff
@@ -28,32 +28,51 @@
+import {injectorBridge} from '../../angular/angular-injector-bridge.functions';
 
 export class SpentTimeDisplayField extends DurationDisplayField {
+  public PathHelper;
```
missing parameter types

romanroe commented on frontend/app/components/wp-display/field-types/wp-display-spent-time-field.module.ts:
```diff
@@ -28,32 +28,51 @@
     };
+  }
+
+  public render(element: HTMLElement, displayText): void {
```
missing parameter types

romanroe commented on frontend/app/components/wp-display/field-types/wp-display-spent-time-field.module.ts:
```diff
@@ -28,32 +28,51 @@
-      resource.project.$load().then(project => {
-        this.timeEntriesLink = URI(this.PathHelper.projectTimeEntriesPath(project.identifier))
-          .search({ work_package_id: resource.id })
+    let link = document.createElement('a');
```
`const` ?

romanroe commented on frontend/app/components/wp-display/field-types/wp-display-work-package-field.module.ts:
```diff
@@ -28,22 +28,39 @@
 export class WorkPackageDisplayField extends DisplayField {
-  public template: string = '/components/wp-display/field-types/wp-display-work-package-field.directive.html';
-  public text: Object;
+  public text:any;
```
really `any`?

romanroe commented on frontend/app/components/wp-display/field-types/wp-display-work-package-field.module.ts:
```diff
@@ -28,22 +28,39 @@
-  public template: string = '/components/wp-display/field-types/wp-display-work-package-field.directive.html';
-  public text: Object;
+  public text:any;
+  private uiStateBuilder;
```
missing parameter types

romanroe commented on frontend/app/components/wp-display/wp-display-field/wp-display-field.module.ts:
```diff
@@ -51,29 +50,47 @@ export class DisplayField extends Field {
+           this.name;
+  }
+
+  public render(element: HTMLElement, displayText): void {
```
missing parameter types

romanroe commented on frontend/app/components/wp-edit-form/table-row-edit-context.ts:
```diff
@@ -0,0 +1,85 @@
+    return jQuery(this.rowElement).find(`.${tdClassName}.${fieldName}`);
+  }
+
+  public reset(workPackage, fieldName:string) {
```
missing parameter types

romanroe commented on frontend/app/components/wp-edit-form/work-package-edit-context.ts:
```diff
@@ -1,12 +1,12 @@
+// -- copyright
 // OpenProject is a project management system.
-// Copyright (C) 2012-2017 the OpenProject Foundation (OPF)
+// Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
```
?

oliverguenther replied:
That's caused by dev having updated the copyright notices in the meantime. Fixed

romanroe commented on frontend/app/components/wp-edit-form/work-package-edit-field-handler.ts:
```diff
@@ -0,0 +1,171 @@
+  /**
+   * Stop this event from propagating out of the edit field context.
+   */
+  public stopPropagation(evt) {
```
missing parameter types

romanroe commented on frontend/app/components/wp-edit-form/work-package-edit-form.ts:
```diff
@@ -0,0 +1,207 @@
+  public editContext:WorkPackageEditContext | null;
+
+  // Other fields
+  public workPackage;
```
missing parameter types

romanroe commented on frontend/app/components/wp-edit-form/work-package-edit-form.ts:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    this.workPackage.loadFormSchema().then(schema => {
+      let field = <EditField> this.wpEditField.getField(
```
... as ....

romanroe commented on frontend/app/components/wp-edit-form/work-package-edit-form.ts:
```diff
@@ -0,0 +1,207 @@
+    });
+  }
+
+ public submit() {
```
missing return type

romanroe commented on frontend/app/components/wp-fast-table/builders/cell-builder.ts:
```diff
@@ -0,0 +1,71 @@
+      return td;
+    }
+
+    const field = <DisplayField> this.wpDisplayField.getField(workPackage, name, fieldSchema);
```
... as ...

romanroe commented on frontend/app/components/wp-fast-table/builders/rows/rows-builder.ts:
```diff
@@ -0,0 +1,44 @@
+export abstract class RowsBuilder {
+  public states:States;
+
+  protected rowBuilder;
```
missing parameter types

romanroe commented on frontend/app/components/wp-fast-table/handlers/cell/edit-cell-handler.ts:
```diff
@@ -0,0 +1,76 @@
+    injectorBridge(this);
+  }
+
+  protected processEvent(table: WorkPackageTable, evt:JQueryEventObject) {
```
1/3 no return type

romanroe commented on frontend/app/components/wp-fast-table/handlers/cell/edit-cell-handler.ts:
```diff
@@ -0,0 +1,76 @@
+    if (!fieldName) {
+      console.warn('Click handled by cell not a field?');
+      console.warn(target);
+      return;
```
2/3 return undefined

romanroe commented on frontend/app/components/wp-fast-table/handlers/cell/edit-cell-handler.ts:
```diff
@@ -0,0 +1,76 @@
+    // Activate the field
+    form.activate(fieldName);
+
+    return false;
```
3/3 return false

=> intention unclear

romanroe commented on frontend/app/components/wp-fast-table/handlers/row/group-row-handler.ts:
```diff
@@ -0,0 +1,55 @@
+    this.collapsedState.put(state);
+
+    // Refresh groups
+    setTimeout(() => {
```
timeout required?

romanroe commented on frontend/app/components/wp-fast-table/state/wp-table-metadata.service.ts:
```diff
@@ -0,0 +1,46 @@
+   * of groupable columns.
+   */
+  public isGroupable(name:string):boolean {
+    return !!_.find(this.current.groupableColumns, (column) => column.name === name);
```
_.find => _.contains ??

oliverguenther replied:
Contains works only for properties. While it appears to look in all values for an object, I find that to be less precise

oliverguenther committed:
Review feedback

oliverguenther committed:
Properly close fields when submitting without changes

oliverguenther committed:
Fixes after last dev rebase

oliverguenther committed:
Fix error in grouping modal

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/dae113ae91f61f1410cc7fbb1f84f822c1bc1cd3) (deployed commit: 5c7a42bd5ce701e5cbb2746a5a9ec90f916de139).

romanroe commented on frontend/app/components/context-menus/context-menu.service.ts:
```diff
@@ -54,21 +57,21 @@ export class ContextMenuService {
+  public close(disableFocus:boolean = false):Promise<void> {
     if (!this.active) {
-      return this.$q.when(true);
+      return this.$q.when(void 0);
```
Used be a `true`, now `void 0`. Are you sure? Also, I find `void 0` uncommon. Why not `null` or `undefined`?

oliverguenther replied:
Its expected to be undefined. Let me know if you want `void(0)` to be removed.

romanroe replied:
I know, I am nitpicking here :-) I simply find `void 0` to be uncommon, at least in our code base. Since you can't override `undefined` anyway, I always use `undefined`. Also, TypeScript's `strictNullChecks` check for `undefined`. Not sure it understands `void 0`.

oliverguenther replied:
Agreed, its removed!

oliverguenther committed:
Replace void(0) with undefined

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/17cbd5a4f850572d3f43a2d8c45122b9cba59f62) (deployed commit: ea3c755530efa78eae8cfd2b550d6d6572ce737f).
